backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,336 @@
+                        existingBackendIPAddress, ProtoUtil.processGroupCompactRepr(processGroup)));
+                    future.complete(existingBackendIPAddress);
+                  } else {
+                    try {
+                      deAssociateBackendWithProcessGroup(existingBackend, processGroup);
+                      associateBackendWithProcessGroup(newBackend, processGroup);
+                      logger.debug(String.format("process_group=%s, de-associating existing backend=%s, associating new backend=%s",
+                          ProtoUtil.processGroupCompactRepr(processGroup), existingBackend.getBackendIPAddress(), newBackend.getBackendIPAddress()));
+                      future.complete(newBackend.getBackendIPAddress());
+                    } catch (Exception ex) {
+                      future.fail(String.format("Cannot persist association of backend=%s with process_group=%s in zookeeper",
+                          newBackend.getBackendIPAddress(), processGroup));
+                    }
+                  }
+                } finally {
+                  if(newBackend != null) {
+                    availableBackendsByPriority.offer(newBackend);
+                  }
+                }
+              }
+            }
+          } else {
+            future.fail("Timeout while acquiring lock on backend queue for process_group=" + processGroup);
+          }
+        } catch (InterruptedException ex) {
+          future.fail(new RuntimeException("Interrupted while acquiring lock on backend queue for process_group=" + processGroup, ex));
+        } finally {
+          backendAssignmentLock.unlock();
+        }
+      }
+    }, result.completer());
+
+    return result;
+  }
+
+  /**
+   * Finds an available backend from the queue. De-queues all backends which have become defunct since they were added
+   * @return available backend or null if none found
+   */
+  private BackendDetail getAvailableBackendFromPriorityQueue() {
+    BackendDetail availableBackend;
+    while((availableBackend = availableBackendsByPriority.poll()) != null) {
+      if(!availableBackend.isDefunct()) {
+        return availableBackend;
+      }
+    }
+    return (availableBackend == null || availableBackend.isDefunct()) ? null : availableBackend;
+  }
+
+  /**
+   * Associates a backend with process group. Writes to in-memory store and zookeeper
+   * NOTE: Ensure that the backend detail instance being modified has been removed from available backend queue prior to calling this method
+   * @param backendDetail
+   * @param processGroup
+   * @throws Exception
+   */
+  private void associateBackendWithProcessGroup(BackendDetail backendDetail, Recorder.ProcessGroup processGroup)
+      throws Exception {
+    Set<Recorder.ProcessGroup> associatedProcessGroups = new HashSet<>(backendDetail.getAssociatedProcessGroups());
```

Comment:
If write to ZK fails, backenddetail held in-memory should not have the association. Therefore, creating a temporary processgroups set for serialization purposes (otherwise will have to add a try catch and rollback association if write to zk fails)